src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -225,14 +197,130 @@ void BaseCouplingScheme::finalize()
+  _time          = startTime;
+  _timeWindows   = startTimeWindow;
+
+  if (_couplingMode == Implicit) {
+    initializeImplicit();
+
+    requireAction(constants::actionWriteIterationCheckpoint());
+    initializeTXTWriters();
+  }
+
+  initializeImplementation();
+
+  if (hasToSendInitData()) {
+    requireAction(constants::actionWriteInitialData());
+  }
+
+  _initializeHasBeenCalled = true;
+}
+
+void BaseCouplingScheme::initializeSendingParticipants(DataMap &dataMap)
+{
+  for (DataMap::value_type &pair : dataMap) {
+    if (pair.second->initialize) {
+      _hasToSendInitData = true;
+      break;
+    }
+  }
+}
+
+void BaseCouplingScheme::initializeReceivingParticipants(DataMap &dataMap)
+{
+  {
+    for (DataMap::value_type &pair : dataMap) {
+      if (pair.second->initialize) {
+        _hasToReceiveInitData = true;
+        break;
+      }
+    }
+  }
+}
+
+
+void BaseCouplingScheme::initializeData()
+{
+  // InitializeData uses the template method pattern (https://en.wikipedia.org/wiki/Template_method_pattern).
+  PRECICE_CHECK(_initializeHasBeenCalled, "initializeData() can be called after initialize() only.");
+  PRECICE_CHECK(not _initializeDataHasBeenCalled, "initializeData() can only be called once.");
+  _initializeDataHasBeenCalled = true;
+  PRECICE_TRACE("initializeData()");
+
+  if (not _hasToSendInitData && not _hasToReceiveInitData) {
+    PRECICE_INFO("initializeData is skipped since no data has to be initialized.");
+    return;
+  }
+
+  PRECICE_DEBUG("Initializing Data ...");
+
+  PRECICE_CHECK(not(_hasToSendInitData && isActionRequired(constants::actionWriteInitialData())),
+                "InitialData has to be written to preCICE before calling initializeData().");
+
+  _hasDataBeenExchanged = false;
+
+  initializeDataImpl();
+}
+
+void BaseCouplingScheme::advance()
+{
+  PRECICE_TRACE(getTimeWindows(), getTime());
+  checkCompletenessRequiredActions();
+  PRECICE_CHECK(_initializeHasBeenCalled, "Before calling advance() coupling has to be initialized via initialize(). This will cause an error in future releases.")  // TODO: preCICE v3.0.0 -> PRECICE_CHECK
+  PRECICE_CHECK((not _hasToReceiveInitData && not _hasToSendInitData) || (_initializeDataHasBeenCalled),
+                "initializeData() needs to be called before advance if data has to be initialized.");
+  _hasDataBeenExchanged = false;
+  _isTimeWindowComplete = false;
+
+#ifndef NDEBUG
+  for (const DataMap::value_type &pair : getReceiveData()) {
+    Eigen::VectorXd &  values = *pair.second->values;
+    int                max    = values.size();
+    std::ostringstream stream;
+    for (int i = 0; (i < max) && (i < 10); i++) {
+      stream << values[i] << " ";
+    }
+    PRECICE_DEBUG("Begin advance, first New Values: " << stream.str());
+  }
+#endif
+  PRECICE_CHECK(_couplingMode != Undefined, "_couplingMode has to be defined.");
+
+  if (subcyclingIsCompleted()) {
+    if (isExplicitCouplingScheme()) {
+      timeWindowCompleted();
```

Comment:
What about "completeWindow"?

Reply:
Or `finalizeWindow`? Like already mentioned somewhere else. `complete` is always a bit unspecific...